rush-plugins/rush-build-graph-plugin/src/GraphParser.ts

```diff
@@ -0,0 +1,197 @@
+        isValid = false;
+      }
+
+      if (!entry.command) {
+        this._logger.terminal.writeErrorLine(`There is an empty command in ${entry.id} `);
+        isValid = false;
+      }
+    }
+    if (isValid) {
+      this._logger.terminal.writeLine(
+        Colorize.green('All nodes have non-empty commands and dependencies which exist')
+      );
+    }
+    const totalEdges: number = entries.reduce((acc, entry) => acc + entry.dependencies.length, 0);
+    this._logger.terminal.writeLine(`Graph has ${entries.length} nodes, ${totalEdges} edges`);
+    return isValid;
+  }
+
+  /*
+   * Get the operation id, throw an error if it fails
+   */
+  public getOperationId(operation: Operation): string {
+    const result: string | undefined = tryGetOperationId(operation);
+    if (!result) {
+      throw new Error(
+        'Operation does not have a name: ' + JSON.stringify(filterObjectForDebug(operation, 2), undefined, 2)
+      );
+    }
+    return result;
+  }
+
+  /*
+   * remove all entries with empty commands
+   * if an entry has a dependency with an empty command, it should inherit the dependencies of the empty command
+   */
+  private _pruneNoOps(inputNodeMap: INodeMapInternal): IGraphNode[] {
+    // Cache for the non-empty upstream dependencies of each operation
+    const nonEmptyDependenciesByOperation: Map<IGraphNodeInternal, Set<string>> = new Map();
+    function getNonEmptyDependencies(node: IGraphNodeInternal): ReadonlySet<string> {
+      // If we've already computed this, return the cached result
+      let nonEmptyDependencies: Set<string> | undefined = nonEmptyDependenciesByOperation.get(node);
+      if (!nonEmptyDependencies) {
+        nonEmptyDependencies = new Set();
+        nonEmptyDependenciesByOperation.set(node, nonEmptyDependencies);
+        for (const dependencyID of node.dependencies) {
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          if (!inputNodeMap.get(dependencyID)!.command) {
+            // If the dependency is empty, recursively inherit its non-empty dependencies
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            for (const deepDependency of getNonEmptyDependencies(inputNodeMap.get(dependencyID)!)) {
+              nonEmptyDependencies.add(deepDependency);
+            }
+          } else {
+            nonEmptyDependencies.add(dependencyID);
+          }
+        }
+      }
+      return nonEmptyDependencies;
+    }
+
+    return Array.from(inputNodeMap.values())
+      .filter((node) => node.command)
+      .map((node) => {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        return { ...node, dependencies: Array.from(getNonEmptyDependencies(node)), command: node.command! };
+      });
+  }
+
+  /*
+   * Convert an operation into a graph node
+   */
+  private _operationAsHashedEntry(operation: Operation): IGraphNodeInternal {
+    const node: Partial<IGraphNodeInternal> = {
```

Comment:
Making this a Partial type is awkward.

Reply:
I think it's appropriate, as I don't want to take it for granted that the parts of a node exist as expected, due to the ability of the rush graph to be tweaked by other plugins.